### PR TITLE
Asset rules engine

### PR DESCRIPTION
**Pull request by wawax**
Add business rules for assets.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

File: inc/commondbtm.class.php [hunk 1]
```diff
@@ -4889,4 +4895,40 @@ public function getAutofillMark($field, $options, $value = null) {
       }
       return $mark;
    }
+
+   /**
+   * Manage business rules for assets
+   *
+   * @since 9.3
+   * @param $add true if an asset is added
```
Comment (trasher) [inline inc/commondbtm.class.php]: missing variable type after @param

File: tests/units/RuleAsset.php [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+   public function testTriggerAdd() {
+      $this->login();
+
+      // prepare rule
+      $this->_createRuleComment(\RuleAsset::ONADD);
+
+      // test create ticket (trigger on title)
+      $computer = new \Computer;
+      $computers_id = $computer->add($computer_input = [
```
Comment (trasher) [inline tests/units/RuleAsset.php]: When adding in the db, test if return is > 0

File: tests/units/RuleAsset.php [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+      ]);
+      $this->string((string)$computer->getField('comment'))->isEqualTo('comment1');
+
+      $computers_id = $computer->add($computer_input = [
+         'name'        => "computer2",
+         'entities_id' => 0,
+         'is_dynamic'  => 0,
+         '_auto'       => 0
+      ]);
+      $computer->getFromDB($computers_id);
```
Comment (trasher) [inline tests/units/RuleAsset.php]: When calling getFromDB, tests is return is true

File: tests/units/RuleAsset.php [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+         '_auto'       => 1
+      ]);
+      $computer->getFromDB($computers_id);
+      //User rule should apply (extract @domain from the name) but should not
+      //find any user, so users_id is set to 0
+      $this->integer((int)$computer->getField('users_id'))->isEqualTo(0);
+
+   }
+/*
+   public function testTriggerUpdate() {
```
Comment (trasher) [inline tests/units/RuleAsset.php]: To drop?

**Comment by wawax** [comment]
@tsmr @ddurieux it's going to impact OCS and FusionInventory.
We're going to replace the computer's user assignation from a hardcoded algorith to a set of rules

**Comment by orthagh** [comment]
before merge, we need to revert #3333 (and migrate if already released)

**Pull request merged**